File: pset1/credit/credit.c
#include <stdio.h>
#include <cs50.h>

//A program to check the brand and validity of a credit card
bool checksum(long ccn);
void printBrand(long ccn);

int main(void)
{
    long creditCardNunmber = get_long("Number: ");
    if (checksum(creditCardNunmber))
    {
        printBrand(creditCardNunmber);
    }
    else
    {
        printf("INVALID\n");
        // return 1;
    }
}

// int ccnLen(long ccn)
// {
//     int len = 0;
//     for (int i = 0; i != '\0'; i++)
//     {
//         len++;
//     }
//     return len;
// }
bool checksum(long ccn)
{
    int i, digit, sum = 0;
    for (i = 0; ccn > 0; ccn /= 10, i++)
    {
        if (i % 2 == 0)
        {
            sum += ccn % 10;
        }
        else
        {
            digit = (ccn % 10) * 2;
            sum += (digit / 10) + (digit % 10); //if digit = 12, / 10 = 1 & % 10 = 2
        }
    }
    // int len = ccnLen(ccn);
    if ((sum % 10 == 0) && (i == 13 || i == 15 || i == 16))
        return true;
    else
        return false;
}

void printBrand(long ccn)
{
    if( (ccn >= 34e13 && ccn < 35e13) || (ccn >= 37e13 && ccn < 38e13) )
        printf("AMEX\n");
    else if (ccn >= 51e14 && ccn < 56e14)
        printf("MASTERCARD\n");
    else if ( (ccn >= 4e12 && ccn < 5e12) || (ccn >= 4e15 && ccn < 5e15) )
        printf("VISA\n");
    else
        printf("INVALID\n");
}

File: pset6/credit/credit.py
from cs50 import get_string



def main():
    while True:
        card_number = get_string("Number: ")
        length = len(card_number)
        # if (length >= 13 and length <= 16) and
        if card_number.isdigit():
            break

    if checksum(card_number) != 0:
        print("INVALID")
    # amex = 15 starts with 34 or 37
    if length == 15 and (card_number[0] == "3" and (card_number[1] == "4" or card_number[1] == "7")):
        print("AMEX")

    # masta = 16 starts with 51 to 55
    elif length == 16 and (card_number[0] == "5"):
        print("MASTERCARD")

    # visa = 13 or 16 starts with 4
    elif length == 13 or length == 16 and (card_number[0] == "4"):
        print("VISA")

#taking number as a string
def checksum(number):
    results = 0
    digit = 0
    #reverse the number or string
    number = number[::-1]
    for i in range(len(number)):
        if i % 2 != 0:
            digit = int(number[i]) * 2
        else:
            digit = int(number[i])
        for num in str(digit):
            results += int(num)
    return results % 2

# taking number as an in
# def checksum1(number):
#     results = 0
#     digit = 0
#     while number > 0:
#         digit = number % 10




# if __name__ == "__name__":
main()

